identity-account-storage/src/storage/stronghold.rs

```diff
@@ -79,138 +87,192 @@ impl Stronghold {
     // Load the chain-specific store
-    let store: Store<'_> = self.store(&fmt_did(did));
-    let data: Option<Vec<u8>> = store.get(location_chain_state()).await?;
+    let store: Store<'_> = self.store(&fmt_account_id(account_id));
+    let data: Option<Vec<u8>> = store.get(CHAIN_STATE_PATH).await?;
 
     match data {
       None => return Ok(None),
       Some(data) => Ok(Some(ChainState::from_json_slice(&data)?)),
     }
   }
 
-  async fn set_chain_state(&self, did: &IotaDID, chain_state: &ChainState) -> Result<()> {
+  async fn set_chain_state(&self, account_id: &AccountId, chain_state: &ChainState) -> Result<()> {
     // Load the chain-specific store
-    let store: Store<'_> = self.store(&fmt_did(did));
+    let store: Store<'_> = self.store(&fmt_account_id(account_id));
 
     let json: Vec<u8> = chain_state.to_json_vec()?;
 
-    store.set(location_chain_state(), json, None).await?;
+    store.set(CHAIN_STATE_PATH, json, None).await?;
 
     Ok(())
   }
 
-  async fn state(&self, did: &IotaDID) -> Result<Option<IdentityState>> {
+  async fn document(&self, account_id: &AccountId) -> Result<Option<IotaDocument>> {
     // Load the chain-specific store
-    let store: Store<'_> = self.store(&fmt_did(did));
+    let store: Store<'_> = self.store(&fmt_account_id(account_id));
 
     // Read the state from the stronghold snapshot
-    let data: Option<Vec<u8>> = store.get(location_state()).await?;
+    let data: Option<Vec<u8>> = store.get(DOCUMENT_PATH).await?;
 
     match data {
       None => return Ok(None),
-      Some(data) => Ok(Some(IdentityState::from_json_slice(&data)?)),
+      Some(data) => Ok(Some(IotaDocument::from_json_slice(&data)?)),
     }
   }
 
-  async fn set_state(&self, did: &IotaDID, state: &IdentityState) -> Result<()> {
+  async fn set_document(&self, account_id: &AccountId, document: &IotaDocument) -> Result<()> {
     // Load the chain-specific store
-    let store: Store<'_> = self.store(&fmt_did(did));
+    let store: Store<'_> = self.store(&fmt_account_id(account_id));
 
     // Serialize the state
-    let json: Vec<u8> = state.to_json_vec()?;
+    let json: Vec<u8> = document.to_json_vec()?;
 
     // Write the state to the stronghold snapshot
-    store.set(location_state(), json, None).await?;
+    store.set(DOCUMENT_PATH, json, None).await?;
+
+    Ok(())
+  }
+
+  async fn purge(&self, did: &IotaDID) -> Result<()> {
+    let account_id: AccountId = match self.index_get(did).await? {
+      Some(account_id) => account_id,
+      None => return Ok(()),
+    };
+
+    // Remove index entry.
+    let store: Store<'_> = self.store(INDEX_PATH);
+    let index_lock: RwLockWriteGuard<'_, _> = self.index_lock.write().await;
+    let mut index: HashMap<IotaDID, AccountId> = get_index(&store).await?;
+    index.remove(did);
+    set_index(&store, index).await?;
+    // Explicitly release the lock early.
+    std::mem::drop(index_lock);
```

Comment:
This might need double-checked locking? (Not sure if that's an anti-pattern). Since a second instance calling `purge` and waiting on the lock would proceed with the deletion since it doesn't know it's been/is being deleted already? Checking the return of `index.remove` would address that edge-case.

Reply:
Thanks for catching that! Added a check on the return value of `index.remove`.